doc/source/data/examples/custom_datasource_example.rst

```diff
@@ -0,0 +1,107 @@
+Read and Write Custom File Types 
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This guide shows you how to extend Ray Data to read and write file types that aren't 
+natively supported. Images are already supported with the :func:`~ray.data.read_images` 
+and :meth:`~ray.data.Dataset.write_images` APIs, but this example shows you how to 
+implement them for illustrative purposes.
+
+Read data from files
+--------------------
+
+The core abstraction for reading files is :class:`~ray.data.datasource.FileBasedDatasource`.
+It provides file-specific functionality on top of the 
+:class:`~ray.data.Datasource` interface.
+
+To subclass :class:`~ray.data.datasource.FileBasedDatasource`, implement the constructor 
+and :meth:`~ray.data.datasource.FileBasedDatasource._read_stream`.
+
+Implement the constructor
+=========================
+
+Call the superclass constructor and specify the files you want to read.
+Optionally, specify valid file extensions. Ray Data ignores files with other extensions.
+
+.. literalinclude:: ../doc_code/custom_datasource_example.py
+    :language: python
+    :start-after: __datasource_constructor_start__
+    :end-before: __datasource_constructor_end__
+
+Implement ``_read_stream``
+==========================
+
+``_read_stream`` is a generator that yields one or more blocks of data from a file. 
+
+Blocks are PyArrow tables, pandas DataFrames, or dictionaries of NumPy arrays. Don't 
```

Comment:
Add source code link (or pyclass link if it's available?) for Block, DelegatingBlockBuilder and -> "Blocks are a Data-internal representation for PyArrow tables, ..."